test/tests/functional/pbs_hook_execjob_end.py

```diff
@@ -0,0 +1,162 @@
+        date_time3 = str1.split(";")[0]
+        self.logger.info(
+            "execjob_end hook executed at: %s,"
+            "exechost_periodic at: %s and execjob_end hook ended at: %s" %
+            (date_time1, date_time2, date_time3))
+
+    def test_execjob_end_hook_order_and_reject(self):
+        """
+        Test with multiple execjob_end hooks having different order
+        with one of the hooks rejecting the job.
+        """
+        hook_name1 = "execjob_end_logmsg1"
+        hook_body = ("import pbs\n"
+                     "e = pbs.event()\n"
+                     "pbs.logjobmsg(e.job.id, \
+                                  'executed %s hook' % e.hook_name)\n"
+                     "e.accept()\n")
+        attr = {'event': 'execjob_end', 'order': '1', 'enabled': 'True'}
+        self.server.create_import_hook(hook_name1, attr, hook_body)
+        hook_name = "execjob_end_logmsg2"
+        hook_body1 = ("import pbs\n"
+                      "e = pbs.event()\n"
+                      "pbs.logjobmsg(e.job.id, 'executed execjob_end hook')\n"
+                      "e.reject('Job is rejected')\n")
+        attr = {'event': 'execjob_end', 'order': '2', 'enabled': 'True'}
+        self.server.create_import_hook(hook_name, attr, hook_body1)
+        hook_name2 = "execjob_end_logmsg3"
+        attr = {'event': 'execjob_end', 'order': '170', 'enabled': 'True'}
+        self.server.create_import_hook(hook_name2, attr, hook_body)
+        j = Job(TEST_USER)
+        j.set_sleep_time(1)
+        jid = self.server.submit(j)
+        self.mom.log_match("Job;%s;executed %s hook" % (jid, hook_name1),
+                           n=100, max_attempts=10, interval=2)
+        self.mom.log_match("Job;%s;Job is rejected" % jid,
+                           n=100, max_attempts=10, interval=2)
+        self.mom.log_match("Job;%s;executed %s hook" % (jid, hook_name2),
+                           n=100, max_attempts=10, interval=2, existence=False)
+
+    def test_execjob_end_multi_job(self):
+        """
+        Test a execjob_end hook for non-blocking of mom with mutiple jobs
+        """
+        hook_name = "execjob_end_logmsg4"
+        self.server.create_import_hook(hook_name, self.attr, self.hook_body)
+        j = Job(TEST_USER)
+        j.set_sleep_time(1)
+        jid1 = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        j = Job(TEST_USER)
+        j.set_sleep_time(1)
+        jid2 = self.server.submit(j)
+        (_, str1) = self.mom.log_match("Job;%s;executed execjob_end hook" %
+                                       jid1, n=100, max_attempts=10,
+                                       interval=2)
+        date_time1 = str1.split(";")[0]
+        epoch1 = int(time.mktime(time.strptime(
+            date_time1, '%m/%d/%Y %H:%M:%S')))
+        (_, str1) = self.mom.log_match("Job;%s;executed execjob_end hook" %
+                                       jid2, starttime=epoch1, n=100,
```

Comment:
we should also add an endtime of epoch1+10 (10 because we are sleeping for 10s in the end job hook) since we really expect this to appear before that. it is quite possible there could be some delay in 2 jobs execution causing second job message to be printed little late.

Reply:
@saimanikant max_attempts=10 interval=2 will wait for 20s. in that case you might want to update interval=10.
Also please add comments in the code explaining what you are trying to verify.

Reply:
@saimanikant I just realized that max_attempts will only check now+10s. it is quite possible that now is not equals to epoch1 but few seconds away. Hence I suggest to update to endtime=epoch1+10. Also increase the hook sleep time to 15s to avoid delays specially hook copy to moms.

Reply:
Agreed! I'll update this to endtime=epoch1+10. Hook execution starts only after the hook is copied to mom. So I think there's no need to increase sleep time.